File: src/main/java/chapter3/FlatMapExample.java
package chapter3;

import io.reactivex.Observable;
import io.reactivex.functions.Function;


public class FlatMapExample {

    public void marbleDiagram() {
        Function<String, Observable<String>> getDoubleDiamonds =
                ball -> Observable.just(ball + "<>", ball + "<>");

        String[] balls = {"1", "3", "5"};
        Observable<String> source = Observable.fromArray(balls)
                .flatMap(getDoubleDiamonds);
        source.subscribe(System.out::println);
    }

    public void flatMapLambda() {
        String[] balls = {"1", "3", "5"};
        Observable<String> source = Observable.fromArray(balls)
                .flatMap(ball -> Observable.just(ball + "<>", ball + "<>"));
        source.subscribe(System.out::println);
    }

    public static void main(String[] args) {
        FlatMapExample demo = new FlatMapExample();
        demo.marbleDiagram();
        demo.flatMapLambda();
    }

}

File: src/main/kotlin/chapter3/FlatMapExample_.kt
package chapter3

import io.reactivex.Observable


class FlatMapExample_ {

    fun marbleDiagram() {
        val getDoubleDiamonds = { ball: String -> Observable.just<String>(ball + "<>", ball + "<>") }

        val balls = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*balls)
                .flatMap { getDoubleDiamonds(it) }
        source.subscribe { println(it) }
    }

    fun flatMapLambda() {
        val balls = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*balls)
                .flatMap { Observable.just(it + "<>", it + "<>") }
        source.subscribe { println(it) }
    }

}


fun main(args: Array<String>) {
    val demo = FlatMapExample_()
    demo.marbleDiagram()
    demo.flatMapLambda()
}
